viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -647,10 +647,31 @@ void SLAMServiceImpl::SaveMapWithTimestamp() {
     }
 }
 
+void SLAMServiceImpl::AddSensorReading(
```

Comment:
1. This function should return a boolean (true for if the lock was acquired & the sensor data was written to cartographer & false if neither of those things happened).
2. This function should `attempt` to aquire the `map_builder_mutex` & give up if it is unable to acquire the lock.
3. I wasn't aware that the `viam_response_mutex` existed when I designed this api. I thought that there was only one mutex. I think we need to know what other code locks `viam_response_mutex` & how long this lock might block for. If it is always going to be nanoseconds I think this implementation is fine but if it could be longer, I think we need to discuss more what we do.

Reply:
I think GetPointCloudMap will paint the entire map, which I think is the most expensive operation possible, isn't it. 

@kkufieta is there a reason why we use a single lock around both updating the map (which is big & slow) and the position (which is small & fast)?

It seems like we could likely gain a lot of performance by having different locks around these two different entities but please let me know if I'm missing something.

Reply:
I believe using two different locks was being considered, but since we're only copying a string or `Rigid3d` object, we assumed those operations are fast enough to be done under one shared mutex - and the code would be easier to maintain that way.

Reply:
Thanks @kkufieta! I've updated my question